### DISCUSSION TITLE
[SOLVED] I need help with my custom lexer

### DISCUSSION THREAD

**Discussion by Allanon71** [orbitalquark/textadept#204] [Q&A]
[SOLVED]

Hello, 
I'm migrating my lexer definition from an old 8.7 to the newest 11.3, I've almost completed all the needed conversions following the guidelines but I'm not able to let TextAdept recognize my class methods and I'm striggling with it :)
The language I'm using is very similar to LUA in its sytntax.

I think the problem is here:
```
local library = token(lexer.CLASS, word_match{
'http:New',
...
})

lex:add_rule(lexer.CLASS, -B('.') * library)
lex:add_style(lexer.CLASS, lexer.styles.class)
```

Using this definition TextAdept is able to recognize and hilight
`http:New`
but not
`http:New()`

![image](https://user-images.githubusercontent.com/30123216/170327520-dcfe9db6-dcd8-480d-b11a-e34856f5da72.png)

I would be infinitely grateful to anyone who could help me, TextAdept is an incredibly powerful and flexible tool, but I'm getting lost in this thing that is perhaps very easy to solve.

P.S.: I've tried to search for similar questions but I've found a solution for my problem.

------------

Solution is here : https://github.com/orbitalquark/textadept/discussions/204#discussioncomment-2833217

**Comment by orbitalquark** [comment]
Without seeing the entire lexer, I cannot pinpoint what's causing what you see. As a test, I copied the Lua lexer and inserted your 'http:New' in Lua's `library` rule and everything worked just fine. It highlighted both `http:New` and `http:New()` as expected. If you don't mind posting your lexer here or in a gist, I can take a look.

**Reply by Allanon71** [reply]
Thank you very much for your help, no problem sharing my lexer file, I've created a gist here -> 

https://gist.github.com/Allanon71/f82dbaaf65ec072ee17f81482f40fd73

Maybe i've messed something trying to migrate from 8.7 to 11.3  :)

**Comment by orbitalquark** [comment]
It's because you have parentheses in some words in your list. Because of that, Textadept considers parentheses to be "word characters" and looks up `http:New()` as a "library" instead of just the `http:New` part. Since there's no match for the whole thing, you don't get the highlighting.

I would recommend sticking to just words in word lists without any punctuation characters. If you need punctuation, you can add them after a word list. For example:

    local library = token(lexer.CLASS, word_match{
      'http:New',
      ...
    } + 'HGui.MinSBCursorSize (in pixels, default' +
        'HGui.Menu_Open({ MenuDef')

**Reply by Allanon71** [reply]
Thank you very much for your time, I understand what you say but I don't understand different behaviours between the old v8 and the newest v11.3 in syntax hilighting, I'm pretty sure I've messed up something since old version was hilighting pretty fine for my needs.
I'm attaching two screenshots, one shows the old version and one the 11.3, the lexer is the same excluding the changes I've done to adapt it to the new guidelines:

8.7
![shot-87](https://user-images.githubusercontent.com/30123216/170664033-92503e72-28af-4b1f-8352-cd49f3cc06de.png)

11.3
![shot-113](https://user-images.githubusercontent.com/30123216/170664107-dd4749a5-afe5-4128-9f31-888d70aec297.png)

As you can see, methods wasn't hilighted but the classes was (`HGui.Window` and` HGui.Gadget`), also `HGui.SetAutoRendering` was hilighted just fine.
During the typing, if I type `HGui.SetAutoRendering` TextAdept hilight it, but remove the hilight as soon as I add the first bracket.

Thank you in advance for your help :)

**Reply by Allanon71** [reply]
...maybe 'word_match' is not what I need, but something like 'starts_with'?

**Reply by Allanon71** [reply]
Hello again :)
I think I've found a way hacking the lexer.lua source and adding a new custom function called `M.my_word_match(word_list, case_insensitive, word_chars)`
Then I replaced my func in the Hollywood lexer file to hilight my library set, here is a screenshot of the simple hack I did. 
Since I'm a really noob at regex and lpeg I'm not sure if this change can be safely used, right now syntax is hilighted just fine:

![image](https://user-images.githubusercontent.com/30123216/170678373-170cf11d-2d32-4ba6-aa17-65d101b20647.png)

I added an `(` in the pattern

**Comment by orbitalquark** [comment]
That's a good idea to use a copy of the old function in your lexer. I'm glad you resolved the issue.

**Reply by Allanon71** [reply]
Thank you for the suggestions, now all is working just fine :)
